### ISSUE TITLE
PR sandboxes security violation

### ISSUE THREAD

**Issue by Gormartsen** [backdrop-ops/backdropcms.org#208]
There is possible situations when PR would contain some code to violate a security:
- send spam
- add virus
- maybe some Trojan code

It's nice to have a sandbox right away but this issues need to be taken care off.

Any suggestions to keep it simple for Nate but do cover this potential security risks?

**Comment by klonos** [comment]
I was about to file an issue for this, but I know that security matters are dealt with in a separate, private queue to which access have only the security team members. @quicksketch, should we be discussing this in private elsewhere and remove this issue here?

**Comment by Gormartsen** [comment]
@klonos - nothing really bad can happen at this time. If somebody start to send spam - it will be blocked . 

Trojan and viruses is possible case, but again it will affect only people who go to check sandbox website only.

**Comment by klonos** [comment]
> ... If somebody start to send spam - it will be blocked

How? What stops somebody logging in as admin, changing the email body of the welcome new user message for example, then creating users with email addresses and sending them the tampered email message? If this went on for a while, I'm sure that the entire backdropcms.org domain would gain bad reputation and end up in RBL lists.

**Comment by Gormartsen** [comment]
Server located in ovh.
When they detect outgoing spam, they block any outgoing mails and I get a notification.
They analyze any outgoing email via SMTP transparent proxy.

It's quite afficient.

Also there is zabbixmonitoring and as soon as there is to many sending emails, my admins will get notification and take action.

There is also a shell activity detection in the server .
When some script start to use she'll functions like access to passwd or doing other unusual for Drupal activities we have notification as well at our company chat.



Regards,
Gor Martsen
@gormartsen

**Comment by klonos** [comment]
Pretty neat. I guess we've covered the spam risk then. :+1:

**Comment by docwilmot** [comment]
But is there any way to protect access to this though? Right now anyone, even anonymous, can access the sandboxes and login as an admin. That don't seem right to me.

**Comment by quicksketch** [comment]
> @quicksketch, should we be discussing this in private elsewhere and remove this issue here?

I think this is fine to discuss here, the only thing at stake is the sandbox server itself, which Gor is hosting at his company. We do have a private repository for discussing security issues in core and contrib modules, and we handle those privately until a fix is available. That process is in place to protect our users, but the sandbox server shouldn't affect anyone outside of ourselves.

> But is there any way to protect access to this though?

I was curious is adding an additional layer of password protection to the entire site would be appropriate, such as an HTTP prompt, even it told you the user name and password in the prompt itself. That would prevent any spiders and nearly all bots from accessing the sandboxes.

**Comment by Gormartsen** [comment]
> I was curious is adding an additional layer of password protection to the entire site would be appropriate, such as an HTTP prompt, even it told you the user name and password in the prompt itself.

We can setup the same admin/PAss for each sandbox for htpasswd.

**Comment by klonos** [comment]
In my user profile on GitHub I see two-factor authentication, OAuth, personal access tokens that can be used to access the GitHub API as well as SSH keys. Any way we could integrate/use any of these as a second layer of authentication?

This could help with #205

**Comment by klonos** [comment]
...at some point I remember a module or infrastructure implemented in d.org to allow a single login for all sub-sites. Was it https://www.drupal.org/project/sso or its successor https://www.drupal.org/project/cas or some other solution that they used?

**Comment by klonos** [comment]
...point is that most people using the sandboxes will probably have GitHub accounts and most likely backdropcms.org accounts too. We could add those as an additional layer of authentication.

**Comment by Gormartsen** [comment]
I was thinking about using github info, but then we limit sandboxes to github users only and make it a little bit harder to see changes.

My point is that security should be just enough to keep sandbox visitors safe from Trojan and viruses.

There is a possibility to check sandbox for viruses and if found, post comment to PR.

Regards,
Gor Martsen
@gormartsen

**Comment by jenlampton** [comment]
@quicksketch @BWPanda have these same security concerns been addressed with the new tugboat sandboxes? Maybe tugboat handles this already, but just wanted to make sure the new sandboxes are as secure as the old ones.

**Comment by BWPanda** [comment]
I have asked the Tugboat guys in Slack about this.

**Comment by BWPanda** [comment]
Ben from Tugboat replied on Slack (replies copied here):

>> BWPanda said: Spam (users logging into a sandbox site and spamming people from it)
>
> We block all common outbound SMTP ports to help prevent this. It’s not foolproof, but someone would have to go out of their way to create (or find) an open SMTP relay on an odd port to get around it
> We block outbound ports: 25, 465, 587, and 2525

>> BWPanda said: Viruses/trojans (I’m not sure how this works...)
>
> There are a few different considerations here. 1. posting malware that could infect visitors of a preview site 2. posting code to infect the backdrop git repository
> (1) is really outside of what Tugboat can help with. Standard website security practices apply I suppose, but the complication is that the admin credentials are publicly known.
> (2) TL;DR; if you don’t add the ssh key from your tugboat repo settings to your git repo, you’re protected from tugboat-specific attacks
> We manage two sets of ssh keys for every tugboat repo
> one is used by Tugboat directly to pull code from git, and is automatically attached to your github/gitlab/bitbucket repo as a deploy key when you set up Tugboat
> once Tugboat is done pulling/merging/etc, that key is scrubbed from the service containers, and the second keypair is added in its place
> This second key is what you see in your repo settings, and is loosely user manageable (it can be regenerated)
> That key is what your build scripts in config.yml use by default if/when you do anything there over ssh
> That means that the public key from the repo settings needs to be manually imported to wherever you are connecting to via ssh in order for those build script commands to work
> If you are not using that key anywhere, then it’s just a key pair sitting on the containers, nothing can be done with it
> So, if someone were able to get shell access to a preview container, they would have access to a git clone, but no way to push anything back upstream

**Comment by oadaeh** [comment]
Is this still a thing?

**Comment by BWPanda** [comment]
I'm not sure this is something we should be worried about. Tugboat would be handling a lot of this on their end, and as for publicly available login credentials, even if we got rid of that somehow, attackers could just spin up a Tugboat demo of Backdrop from our homepage...

**Comment by jenlampton** [comment]
This was more of an issue when we were relying on the ZenCI sandboxes. I think we can close this issue.
